src/goto-programs/goto_convert.cpp

```diff
@@ -347,11 +347,46 @@ void goto_convertt::convert_label(
   // magic thread creation label?
   if(has_prefix(id2string(label), "__CPROVER_ASYNC_"))
   {
-    // this is like a START_THREAD
-    codet tmp_code(ID_start_thread);
-    tmp_code.copy_to_operands(code.op0());
-    tmp_code.add_source_location()=code.source_location();
-    convert(tmp_code, tmp);
+    // A: START_THREAD : C
+    // B: GOTO Z
+    // C  SKIP
+    // D: BLOCK
+    // E: END_THREAD
+    // Z: SKIP
```

Comment:
I'd really like to understand why this is necessary?

Reply:
@tautschnig The use of labels starting with a specific prefix (`__CPROVER_ASYNC_`) is a viable way to insert a goto instruction `START_THREAD` in the goto program only for C programs. In a Java program you can write labels and `goto` them, but those labels get removed by the Java compiler, they are not present in the `.class` file. As a result, this is not a viable way to tell to CBMC where to insert a `START_THREAD`.

What we instead have done to support both C and Java concurrency is as follow:

- In this point of the file, instead of inserting a `codet` `ID_start_thread`, we insert the actual goto instructions that will be inserted when `ID_start_thread` was later parsed (in this file), namely, the pseudocode instructions in the comment above this line.
- In Java, we have exported two new functions in the `org.cprover.CProver` class that enable us to insert the `START_THREAD` and `END_THREAD` at precise locations (PR to come next week). We will parse calls to those two methods in the `CProver` class before the execution reaches `goto_convertt`, and transform them into `codet`s `ID_{start,end}_thread`.
- We have stored in the `codet` for `ID_start_thread` a label to the first instruction of the thread, and `goto_convertt::convert_start_thread()` uses the standard mechanisms in `goto_convertt` to resolve label targets to resolve the target of the `START_THREAD` instruction.

Reply:
Thanks a lot for the info. Really all of that must go in the source code, whether as comments or additional code.

Can you please clarify why you need the `END_THREAD` to be specified in the Java code? To me, that seems to be the only reason that you need to make those changes in the transformation?

Reply:
@tautschnig, previously   'end_thread' , was inserted in same function  where the start_thread instruction  was converted into goto (goto_convertt::convert_start_thread).  This was viable since,  the entire body of the thread was in op0 of the label, and thus the end of a thread is easily determined. Since In java we can not use labels, the only way to determine (in a generic way), where a thread starts and thread ends, is by inserting a some sort of place holder (i.e org.cprover.CProver.startThread() ...... org..cprover.CProver.endThread())

Reply:
Caveat: As I don't know in sufficient detail how you are doing this in the Java front-end I may be making silly comments.

So you generate and `ID_start_thread` code in the Java front-end - what about putting a `code_blockt` into its operand? I believe that block would then nicely capture the end of the thread?

Reply:
Indeed we generate both ID_start_thread and ID_end_thread in the front end. What you are saying would be possible after a quite  involved analysis of  a multi-block codet, we believe that it is  just cleaner  to insert the  aforementioned  IDs using explicit  method-calls  in the cprover class, will discuss this in more depth in the next pr.

Reply:
So are you doing this involved analysis already? Again, I may be naive: you see to be doing `start_thread <code> end_thread`; my suggestion is to do `start_thread { <code> }`.

Reply:
Yes, we the model for `java.lang.Thread.start()` will contain something like:

```
CProver.start_thread()
<code>
CProver.end_thread()
```

It's impossible tu use `start_thread { <code> }` because the Java compiler will just remove the block and no trace of it will be present in the `.class` file.

Reply:
And no, we are not doing the "involved analysis" that Kurt described above. It is not necessary to insert `START_THREAD` and `END_THREAD` in the code snippet of my previous commit, but would be necessary if we were to transform `<code>` (in previous snippet) to an isolated `codet` to pass around. All we do is to substitute the calls to `CProver.{start,end}_thread()` with the appropriate goto instructions, and generate the labels of the `GOTO` instructions for the parent thread using the general label mechanism in `goto_convertt`.

Reply:
I really want to see the Java front-end part to this. Constructing codet with labels that look like those a user may have provided is confusing and possibly dangerous. It will also break dump-c.

Reply:
@tautschnig You need to make the above observation factual, otherwise it's going to be difficult to continue the discussion.

What is confusing in constructing those `codet`? Please be concrete
What is dangerous about it?
What can break in `dump-c`?

Reply:
@tautschnig It seems that you are going over the same thinking process that @Degiorgio and I already went 3 months ago. In order to avoid repeating the same and getting on the same page I propose that we do a telephone call. I will write you a mail.

Reply:
Phone call scheduled; in the meantime: the labels you are using can legitimately occur in source code. Is it unlikely? Yes, probably. But this risk seems unnecessary to take here. Doing `--show-goto-functions` or `--dump-c` will make those labels visible, and thus users may be confused to see something they never put in there in the first place. Running it through `dump-c` will yield code including those labels, and re-compiling the code will yield new labels in addition to the old ones.

I appreciate that you may have gone through all this, but neither the code nor its comments reflected any of those discussions.